streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignor.java

```diff
@@ -182,11 +186,73 @@ public Subscription subscription(Set<String> topics) {
             state.capacity = state.capacity + 1d;
         }
 
-        // get the tasks as partition groups from the partition grouper
+        // ensure the co-partitioning topics within the group have the same number of partitions,
+        // and enforce the number of partitions for those internal topics.
         Map<Integer, Set<String>> sourceTopicGroups = new HashMap<>();
+        Map<Integer, Set<String>> internalSourceTopicGroups = new HashMap<>();
         for (Map.Entry<Integer, TopologyBuilder.TopicsInfo> entry : topicGroups.entrySet()) {
             sourceTopicGroups.put(entry.getKey(), entry.getValue().sourceTopics);
+            internalSourceTopicGroups.put(entry.getKey(), entry.getValue().interSourceTopics);
         }
+        Collection<Set<String>> copartitionTopicGroups = streamThread.builder.copartitionGroups();
+
+        ensureCopartitioning(copartitionTopicGroups, internalSourceTopicGroups, metadata);
+
+        // for those internal source topics that do not have co-partition enforcement,
+        // set the number of partitions to the maximum of the depending sub-topologies source topics
+        for (Map.Entry<Integer, TopologyBuilder.TopicsInfo> entry : topicGroups.entrySet()) {
+            Set<String> internalTopics = entry.getValue().interSourceTopics;
+            for (String internalTopic : internalTopics) {
+                Set<TaskId> tasks = internalSourceTopicToTaskIds.get(internalTopic);
+
+                if (tasks == null) {
+                    int numPartitions = -1;
+                    for (Map.Entry<Integer, TopologyBuilder.TopicsInfo> other : topicGroups.entrySet()) {
+                        Set<String> otherSinkTopics = other.getValue().sinkTopics;
+
+                        if (otherSinkTopics.contains(internalTopic)) {
+                            for (String topic : other.getValue().sourceTopics) {
+                                List<PartitionInfo> infos = metadata.partitionsForTopic(topic);
+
+                                if (infos != null && infos.size() > numPartitions)
+                                    numPartitions = infos.size();
+                            }
+                        }
+                    }
+
+                    internalSourceTopicToTaskIds.put(internalTopic, Collections.singleton(new TaskId(entry.getKey(), numPartitions)));
+                }
+            }
+        }
+
+        // if ZK is specified, prepare the internal source topic before calling partition grouper
+        if (internalTopicManager != null) {
+            log.debug("Starting to validate internal source topics in partition assignor.");
+
+            for (Map.Entry<String, Set<TaskId>> entry : internalSourceTopicToTaskIds.entrySet()) {
```

Comment:
@guozhangwang: When running the integration test of the latest CP examples this line triggered a `NullPointerException` consistently.  Currently testing a bug fix.  Since this bug is in both this TP branch and in OS Kafka `trunk` we'll need a Kafka JIRA to resolve it.